### ISSUE TITLE
"Differ by a constant" option for mathematical expression part marking

### ISSUE THREAD

**Issue by christianp** [numbas/Numbas#572]
When marking an indefinite integral, the student can pick a name for their constant of integration. Their answer is correct if:

* the intersection of the set of variables they used and the set of variables in the expected answer has size one less than the set of expected variables
* the set of values of the student's expression evaluated at randomly chosen points differs from the expected answer's by a constant.

For the "absolute difference" test, you want to check that `max(student-expected) - min(student-expected) < tolerance`. 

For "relative difference", you want `max(student/expected)/min(student/expected) < tolerance`. (plus whatever fiddling to do with division by zero)

**Comment by christianp** [comment]
The "student can pick a name for the constant of integration" bit doesn't work: the variables are given a random value for each evaluation, so since they expressions use different names, the difference between the two expressions isn't a constant. 

So you have to identify the constant of integration, rewrite the student's expression to use the same name as the expected expression, and then check if they differ by a constant.

Assuming the student uses the expected name, this technique would pick up cases like where the expected answer is `x + C`, and the student writes `x + 2 + C`.

**Comment by christianp** [comment]
It doesn't catch cases where the constant is scaled, e.g. `x + 2*C` instead of `x + C`. For that, you'd need to make sure `C` really is a constant, taking the same value across all evaluations.

**Comment by sangwinc** [comment]
@anst-i and I wrote a paper about this recently, and it's a very interesting topic!

[2024-ConstInt.pdf](https://github.com/user-attachments/files/27002982/2024-ConstInt.pdf)

**Comment by christianp** [comment]
Yes, I think we've talked about it in the past. Thanks for the link to the paper!

**Comment by anst-i** [comment]
In STACK we now also have two independent answer tests to check for the answer (a) being an anti-derivative and (b) having an additive constant. For the second part the default option is to do a strict check only for an actual `+C` (where a `+2` would be considered part of the antiderivative), and a `NONSTRICT` option which accepts any additive term containing a constant. This is mathematically not sound, since checking algorithmically for surjectivity is very hard, so something semi-correct like `+C^2` is accepted.

**Comment by christianp** [comment]
Yes, I think that's how we've been marking integrals in practice: check the derivative.

This issue is almost 8 years old so I'm going to close it, because I don't think that checking expressions differ by a constant is worth the effort.
